Framework/API/src/ExperimentInfo.cpp

```diff
@@ -414,46 +434,93 @@ void ExperimentInfo::swapInstrumentParameters(Geometry::ParameterMap &pmap) {
  */
 void ExperimentInfo::cacheDetectorGroupings(const det2group_map &mapping) {
   populateIfNotLoaded();
-  m_detgroups.clear();
+  setNumberOfDetectorGroups(mapping.size());
+  size_t specIndex = 0;
   for (const auto &item : mapping) {
-    m_det2group[item.first] = m_detgroups.size();
-    m_detgroups.push_back(item.second);
+    m_det2group[item.first] = specIndex;
+    updateCachedDetectorGrouping(specIndex, item.second);
+    specIndex++;
   }
   // Create default grouping if `mapping` is empty.
   cacheDefaultDetectorGrouping();
 }
 
-/// Returns the detector IDs that make up the group that this ID is part of
-const std::set<detid_t> &
-ExperimentInfo::getGroupMembers(const detid_t detID) const {
+/** Sets the number of detector groups.
+ *
+ * This method should not need to be called explicitly. The number of detector
+ * groups will be set either when initializing a MatrixWorkspace, or by calling
+ * `cacheDetectorGroupings` for an ExperimentInfo stored in an MDWorkspace. */
+void ExperimentInfo::setNumberOfDetectorGroups(const size_t count) {
   populateIfNotLoaded();
-  auto iter = m_det2group.find(detID);
-  if (iter != m_det2group.end()) {
-    return m_detgroups[iter->second];
-  } else {
-    throw std::runtime_error(
-        "ExperimentInfo::getGroupMembers - Unable to find ID " +
-        std::to_string(detID) + " in lookup");
+  m_spectrumInfo = Kernel::make_unique<Beamline::SpectrumInfo>(count);
```

Comment:
From what I can tell, at the point this is called you always need a new `m_spectrumInfo` so it would be wasteful to do a `count == m_spectrumInfo.size()` type check prior to the assignment?